0022-compartmentalize-grains-via-pop.md

```diff
@@ -0,0 +1,74 @@
+# Summary
+[summary]: #summary
+
+Grains is currently a static component of salt.  We want it to be developed in a more compartmentalize fashion.
+The engineers at SaltStack have created platform-specific projects that functionally replace the grains
+provided by salt with POP.  We propose that this project, codenamed `grainsv2` becomes available in salt while
+maintaining backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+"The original implementation of grains was not designed to scale to the level that salt requires" - thatch45
+Tom wrote POP to solve this problem and we're going to use it. 
+
+By using Plugin Oriented Programming, grains can be separated into individual projects that can be maintained
+by the community safely.
+
+This also solves the issue of `core.py` being an unmaintainable monolith.  It allows platform specific grains
+to be independently maintained.
+
+Working group leaders and other community members focused on specific platforms can manage the release cycle for
+individual idem-platform projects independently from salt.
+
+# Design
+[design]: #detailed-design
+
+Grains are currently created by calling functions in the salt loader.  We would simply need to call
+external POP functions to collect grains.  We need to maintain backwards compatibility with custom user grains.
+
+- All salt grains have already been ported to `grainsv2`.
+- `grainsv2` has many more features than the salt-grains implementation, but is compatible with salt grains.
+- None of the grain names have been changed, but the output of grains is more consistent/complete between platforms.
+- This will resolve many bugs that are open against grains and will migrate those bugs OUT of the salt
+platform -- making salt easier to develop
+- Grains in the new platform are independently fully tested.
+- Development will freeze on core.py and the salt/grains folder.
+- This comprises a complete rewrite of ALL grains.  Backwards compatibility is expected, but we need this to get
+in early so that any unintentional issues can be reet up a conversation with the relevant parties solved quickly.
+- Salt releases will aggressively pin `grainsv2`/idem-platform dependencies with each release.
+- Users will be to update their grains implementation to a custom idem-platform version.
+- The use of `grainsv2` will be gated with a config variable; It will be disabled by default.
+- New development for grains will be done in `grainsv2` projects and core.py will remain static.
+- When `grainsv2` gains maturity in salt on the other side of the gate (and python3.5 is dropped) it will completely replace core.py
+
+## Alternatives
+[alternatives]: #alternatives
+
+Keep things the way they are.  Does this mean we would suffer in stagnation?  Will we be stuck maintaining a monolith.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What will further adoption of POP mean inside of salt?  The purpose of this sep and these features is not to provide
```

Comment:
While the purpose of this sep does not intend to provide an answer to this question it certainly seems to provide an opinion about it. I feel it would be pertinent to this discussion, especially because the current PR implementing this change is adding a lot more than just a grains specific library. The addition of idem-* means *would be* adding idem and POP into salt core.

Reply:
You're right, this SEP and the accompanying PR do provide strong opinions on this, I mean to say that the future of idem in POP in salt has been discussed by a small group, but warrants a larger discussion, more firm plans, and possibly it's own SEP.   With this [PR](https://github.com/saltstack/salt/pull/57681) we add the `__hub__` to the loader, which means custom execution modules and states could make use of idem's hub... but we have discussed adding the option to specify the salt/idem engine in salt states (defaulting to the salt engine)  is a possible future integration of idem into salt.  This SEP opens the door for that discussion, but that is an implementation/discussion that should happen separately from grains.

Reply:
Seems like the addition of `__hub__` is inviting a lot of confusion and potential complexity to and already complex code base in regards to the loader. What problem is the addition of `__hub__` intended to solve? How will developers know which one to use? What happens if we mix the use of `__salt__` and `__hub__`?